### PR TITLE
Run 2, Run3, and Phase2 tracker: adding T27 (geometry with bricked pixels) and support for storing bricked pixel information

### PR DESCRIPTION
**Pull request by adewit**
#### PR description:

This adds a phase 2 tracker geometry (T27) containing bricked pixels in TEPX&TFPX, as well as the central 1 (3) rod(s) of TBPX L2 (L3 and 4). Because the information about whether a pixel is "bricked" or not should be available at the level of the geometry, a new record to store this information has also been added, as discussed in [these slides](https://indico.cern.ch/event/1045483/contributions/4392063/attachments/2258285/3832430/Brickedness_SIM_AdeWit_0406.pdf). Note that wrt that presentation, the record has been renamed `PTrackerAdditionalParametersPerDetRcd`, and it has been made expandable to make it possible to add further bool/int/float per-detID information for the tracker in the future.
 For now the record is created from the XML, as no Phase 2 geometries are stored in the DB, but a DB writer is also included for future use. 

**NOTE: the new record `PTrackerAdditionalParametersPerDetRcd` is consumed by the TrackerGeomBuilderFromGeometricDet, which is the same for Runs 2&3 and Phase 2. Therefore this PR also modifies Runs 2&3, by including the new record, which however only contains the default value (False) for the brickedness boolean, as this parameter of course does not exist before Phase 2**

The new tracker geometry T27 that uses the bricked pixels is part of detector D85. The workflows for D85 only run up to and including the digitization step to avoid having to include changes in the reconstruction code at this point. **NOTE: the modifications to the digitzer that are needed to handle bricked pixels are NOT included in this PR** ( @franzglessgen will provide that in due course). In the current state the standard digitizer for planar pixels will be used; this is of course not the final configuration but it does not lead to problems.

@emiglior @mmusich fyi

#### PR validation:

- Checked that the addition of the new rcd does not break non-Phase2 workflows (checked for a few workflows)
- Checked that the information on the brickedness of the pixels is passed on as expected
- Verified that the DB writer is working

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Not a backport

### CODE REVIEW

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
> ---> test testTauEmbeddingProducers had ERRORS

this is spurious and due to issue https://github.com/cms-sw/cmssw/issues/34112.
All the rest of failures are real.

**Comment by adewit** [comment]
> > ---> test testTauEmbeddingProducers had ERRORS
> 
> this is spurious and due to issue #34112.
> All the rest of failures are real.

Thanks - I understand all the rest of them apart from the GeometryDOMCount one, which works locally :/

File: HLTrigger/Configuration/test/OnLine_HLT_PIon.py
Comment (mmusich) [inline HLTrigger/Configuration/test/OnLine_HLT_PIon.py:4004]: @adewit, direct modification of the HLT menu are disallowed as per TSG policy. 
Please add a customization function in [customizeHLTforCMSSW](https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/customizeHLTforCMSSW.py)

**Review by mmusich — CHANGES_REQUESTED** [review]

**Comment by adewit** [comment]
> ---> test GeometryDOMCount had ERRORS

After investigation I don't think this one is related to this PR; the IB `CMSSW_12_0_X_2021-06-14-1100` is missing the package `DetectorDescription/Schema`, and that's why this test fails.

File: CondTools/Geometry/plugins/PTrackerAdditionalParametersPerDetDBBuilder.cc
Comment (adewit) [reply CondTools/Geometry/plugins/PTrackerAdditionalParametersPerDetDBBuilder.cc]: Done

File: HLTrigger/Configuration/test/OnLine_HLT_PIon.py
Comment (adewit) [reply HLTrigger/Configuration/test/OnLine_HLT_PIon.py]: Done

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
> ---> test GeometryDOMCount had ERRORS

this should be fixed by this https://github.com/cms-sw/cmssw/pull/34110

> ---> test testTauEmbeddingProducers had ERRORS

this should be fixed by this https://github.com/cms-sw/cmssw/pull/34113

Let's wait for next IB to be available to re-trigger tests.

**Comment by mmusich** [comment]
@cmsbuild, please test

`CMSSW_12_0_X_2021-06-16-0900` seems to be available

File: Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc:423]: This can be now migrated to `esConsumes`

**Review by mmusich — CHANGES_REQUESTED** [review]
@adewit there is a conflict now due to https://github.com/cms-sw/cmssw/pull/33583 recently being merged.
Can you please rebase and in doing so also address the esConsumes in `Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc`?

**Comment by adewit** [comment]
I guess we need to wait for the next IB to do these checks again.

**Comment by mmusich** [comment]
code-checks

**Comment by mmusich** [comment]
test parameters:
   - workflow = 38234.0

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by adewit** [comment]
> I found errors in the following unit tests:
> 
> ---> test test_PixelBaryCentreTool had ERRORS

@mmusich this is the same one that's discussed here: https://github.com/cms-sw/cmssw/pull/33583#issuecomment-866575584, ie already failing in the IB, right?

**Comment by mmusich** [comment]
> ie already failing in the IB, right?

yes

**Comment by mmusich** [comment]
@cms-sw/alca-l2 @cms-sw/db-l2 @cms-sw/geometry-l2 @cms-sw/pdmv-l2 @cms-sw/upgrade-l2 @cms-sw/hlt-l2 @cms-sw/operations-l2 
can you please review this PR at earliest convenience?
The failure in the unit test is already present in the IB and it will take some time to fix it, so please disregard than for the time being.
Thank you

File: CondTools/Geometry/plugins/PTrackerAdditionalParametersPerDetDBBuilder.cc
Comment (cvuosalo) [inline CondTools/Geometry/plugins/PTrackerAdditionalParametersPerDetDBBuilder.cc:9]: These two `DDCompactView` includes seem unnecessary. Please remove them if possible.

**Comment by cvuosalo** [comment]
@adewit Sorry, I am not understanding something. Why are Run 2 and Run 3 files being changed for a new Phase 2 record? What function will `PTrackerAdditionalParametersPerDetRcd` have in Runs 2 and 3?

File: CondTools/Geometry/plugins/PTrackerAdditionalParametersPerDetDBBuilder.cc
Comment (adewit) [reply CondTools/Geometry/plugins/PTrackerAdditionalParametersPerDetDBBuilder.cc]: Indeed! I've removed them.

**Comment by adewit** [comment]
> @adewit Sorry, I am not understanding something. Why are Run 2 and Run 3 files being changed for a new Phase 2 record? What function will `PTrackerAdditionalParametersPerDetRcd` have in Runs 2 and 3?

@cvuosalo for Runs 2 and 3 it indeed has no function at the moment (it'll be created but it'll just be map of detId:bool where the bool is always false). The reason for this is that this record is needed by `TrackerGeomBuilderFromGeometricDet` (https://github.com/cms-sw/cmssw/pull/34120/files#diff-aeb1df46ab2dad00dda0ee9c96c2a30afb705d47aa00aa5308d4673af76b71d7), and there is only one version of this which is used both for Phase 2 and Run 2/3.

File: Alignment/CommonAlignmentMonitor/plugins/AlignmentMonitorAsAnalyzer.cc
Comment (mmusich) [inline Alignment/CommonAlignmentMonitor/plugins/AlignmentMonitorAsAnalyzer.cc:144]: @adewit now that PR https://github.com/cms-sw/cmssw/pull/34270 is merged, this file has conflicts.
Can you please resolve them, and while you are at it use `esConsumes` for the newly introduced record?
Thanks.

**Review by mmusich — APPROVED** [review]

File: Alignment/CommonAlignmentMonitor/plugins/AlignmentMonitorAsAnalyzer.cc
Comment (adewit) [reply Alignment/CommonAlignmentMonitor/plugins/AlignmentMonitorAsAnalyzer.cc]: Done

**Comment by mmusich** [comment]
@adewit finally now there are other conflicts due to PR https://github.com/cms-sw/cmssw/pull/34248.
As usual please migrate to esConsumes when touching the files.
That's the last one (promised :) )

**Comment by adewit** [comment]
> @adewit finally now there are other conflicts due to PR #34248.
> As usual please migrate to esConsumes when touching the files.
> That's the last one (promised :) )

Heh, that'll teach me to do things first thing on Monday before checking the progress of other PR's ;-) 
Anyway, I think we're there now :)

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by tvami** [comment]
+alca

* the unit test failure is known and framework experts are working on it
* otherwise ok for AlCa

**Comment by mmusich** [comment]
@cmsbuild, please test with https://github.com/cms-sw/cmssw/pull/34363

**Comment by kpedro88** [comment]
@adewit it may be useful to squash this branch before the PR is merged, to avoid many added-then-deleted files and reduce the overall size of the content added to the official repository (see https://github.com/cms-sw/cmssw/pull/34120#issuecomment-874089854)

**Comment by tvami** [comment]
@cmsbuild, please test with #34363

**Comment by tvami** [comment]
@cmsbuild, please test with #34363

**Comment by cvuosalo** [comment]
@adewit There's something I don't understand about this implementation. The `isBricked` boolean is set as a `SpecPar` in `pixelStructureTopology.xml`. For geometries that are in the Conditions DB, the `SpecPars` are included. That means the value of being bricked is automatically stored in the DB. Why is the special new `PTrackerAdditionalParametersPerDetRcd` record required?

**Comment by adewit** [comment]
> @adewit There's something I don't understand about this implementation. The `isBricked` boolean is set as a `SpecPar` in `pixelStructureTopology.xml`. For geometries that are in the Conditions DB, the `SpecPars` are included. That means the value of being bricked is automatically stored in the DB. Why is the special new `PTrackerAdditionalParametersPerDetRcd` record required?

@cvuosalo sorry for the late reply. For constructing the tracker geometry from DB we're reading out the `PGeometricDet` record (which can't be modified, so the `isBricked` information can't be added to it), I don't know how this relates to how the geometries themselves are stored in the DB. At the point where the information is read in we need access to both the `detID` (which is not in the geometry xml) and the value of e.g. the `isBricked` bool, and this is where the new record is needed. Again, I don't know how the geometry itself is stored in DB, so I'm presuming here that the `detID`s are not stored along with it, which means not enough information is available. But maybe I missed something.

**Comment by mmusich** [comment]
@cvuosalo, this PR has been open for about 1 month now and its implications shown at the CMS Simulation meeting on June 4th https://indico.cern.ch/event/1045483/#48-phase-2-inner-tracker-brick, with what - at the time - appeared to be general consensus from the geometry group.
Any further delay in the integration will constitute a serious setback in the studies program of the Phase-2 Inner Tracker Task force.
Today @VinInn suggested to an alternative approach by overriding the meaning of certain existing parameters in `PGeometricDet` such as:

https://github.com/cms-sw/cmssw/blob/29db21632a46d8fe3dbc3da684ba0d690880bd3b/CondFormats/GeometryObjects/interface/PGeometricDet.h#L45

which is not used for pixels (and won't be in phase-2)

While this - in my opinion - is a less clean solution (and also less future-proof than the expandable parameter set solution presented here), but would allow to touch perhaps less files (and also won't introduce a new record).
Please clarify at earliest convenience what is the recommended solution from the geometry group.
Thank you

**Comment by cvuosalo** [comment]
This PR seems to run into a number of CMSSW design constraints. It is generally not feasible to expand existing DB records, so we have the choice of misusing an existing unused parameter (`_stereo`) or creating a new DB record. If we add the new record, since our code should not be explicitly run-dependent, the new record has to be included in Runs 2 & 3, even though it will just hold booleans that will always be false for Runs 2 & 3. Finally, the existing design combines in a complex way `detIdShifts` defined by a vector constant in an XML file with `SpecPars`. Since XML constants are not stored in the DB, this combination needs to be stored in the DB. However, there might be an alternate design that would store the `detIdShifts` in `SpecPars`, which are stored in the DB, which would eliminate the need for the new record. But the existing design is well-established, so it is probably best to follow it for consistency rather than try to tack on a new design at this late date.

I don't see good alternatives to the choices in this PR. I wonder about the implications for Runs 2 & 3 if we continue to burden them with unused Phase 2 functionality.

**Comment by cvuosalo** [comment]
@adewit Please change the title of this PR to start "Run 2, Run 3, & Phase 2". It is a big issue that this PR is modifying Run 2, and it needs to be loudly announced. It seems unavoidable, but it must not be hidden. Right now, the naive reader would think this PR only touches Phase 2.
Also, please add in bold letters in the PR description a brief discussion about the changes to Runs 2 & 3.

File: Configuration/Geometry/python/GeometryExtended_cff.py
Comment (cvuosalo) [inline Configuration/Geometry/python/GeometryExtended_cff.py:12]: I would like to a warning to readers about why this line is included. I suggest adding a comment along the lines of:
```
# TrackerAdditionalParametersPerDet contains only default values -- needed for consistency with Phase 2
```
If we have to touch Run 2, at least let's leave some explanation.

File: Configuration/Geometry/python/GeometryRecoTracker_cff.py
Comment (cvuosalo) [inline Configuration/Geometry/python/GeometryRecoTracker_cff.py:9]: Add comment as above please.

File: Configuration/Geometry/python/GeometryReco_cff.py
Comment (cvuosalo) [inline Configuration/Geometry/python/GeometryReco_cff.py:9]: Add comment as above please.

File: Configuration/Geometry/python/GeometryRecoDB_cff.py
Comment (cvuosalo) [inline Configuration/Geometry/python/GeometryRecoDB_cff.py:8]: Again, in these core config files for Runs 2 & 3, please add the comment I mentioned for `GeometryExtended_cff.py`.

**Comment by fwyzard** [comment]
> If we add the new record, since our code should not be explicitly
run-dependent, the new record has to be included in Runs 2 & 3, even though
it will just hold booleans that will always be false for Runs 2 & 3.

Why not use different records for Phase 1 and Phase 2, and let a python
configuration parameter direct which one to use ?

.A

File: Configuration/Geometry/python/GeometryRecoDB_cff.py
Comment (adewit) [reply Configuration/Geometry/python/GeometryRecoDB_cff.py:9]: I've added the comments in the 4 files where they were requested (+ changed the PR title/description).

File: CondFormats/GeometryObjects/src/PTrackerAdditionalParametersPerDet.cc
Comment (makortel) [inline CondFormats/GeometryObjects/src/PTrackerAdditionalParametersPerDet.cc:25]: Sorry to come in late, but it would be better to move these to anonymous namespace to avoid adding these somewhat generic (although templated) names into the global namespace.

**Comment by makortel** [comment]
> If we add the new record, since our code should not be explicitly run-dependent, the new record has to be included in Runs 2 & 3, even though it will just hold booleans that will always be false for Runs 2 & 3.

I understood from the [slides](https://indico.cern.ch/event/1045483/contributions/4392063/attachments/2258285/3832430/Brickedness_SIM_AdeWit_0406.pdf) that nothing would be added to the DB for Runs <= 3, and the "all-false" ESProduct would be created on the fly for those (as in this PR).

(but I may have gotten confused by the use of word "Record" for what looks to me as "data product" or "ESProduct" or "payload"; "Record" really means the entity that defines the IOVs for the data products in that Record)

**Comment by mmusich** [comment]
> I understood from the slides that nothing would be added to the DB for Runs <= 3, and the "all-false" ESProduct would be created on the fly for those (as in this PR).

That's correct and even for phase-2 now it will be created on the fly from the XML descriptions. DB will be populated only for Phase-2 once it will be a "real production" workflow

File: CondFormats/GeometryObjects/src/PTrackerAdditionalParametersPerDet.cc
Comment (adewit) [reply CondFormats/GeometryObjects/src/PTrackerAdditionalParametersPerDet.cc]: I've wrapped them in `namespace {}` which I think does what you asked for, but if I misunderstood please let me know :-)

**Comment by srimanob** [comment]
Hi @adewit 
As asking to you in
https://github.com/cms-sw/cmssw/pull/34481
Would you mind updating the PR to D87 ?

FYI @bsunanda

File: CondFormats/GeometryObjects/src/PTrackerAdditionalParametersPerDet.cc
Comment (makortel) [reply CondFormats/GeometryObjects/src/PTrackerAdditionalParametersPerDet.cc]: That was exactly it, thanks!

**Comment by adewit** [comment]
> Hi @adewit
> As asking to you in
> #34481
> Would you mind updating the PR to D87 ?
> 

@srimanob sorry, I missed your message on the other PR. Yes, will update this one to D87. As it looks like #34481 will be signed off soon it's easier to move after that one is merged (either way I need to update this PR after #34481 is merged, as the PR adds a new import statement in all `GeometryExtended20216...Reco_cff.py`'s, so I figure it's better to do it all in one go)

**Comment by srimanob** [comment]
> > Hi @adewit
> > As asking to you in
> > #34481
> > Would you mind updating the PR to D87 ?
> 
> @srimanob sorry, I missed your message on the other PR. Yes, will update this one to D87. As it looks like #34481 will be signed off soon it's easier to move after that one is merged (either way I need to update this PR after #34481 is merged, as the PR adds a new import statement in all `GeometryExtended20216...Reco_cff.py`'s, so I figure it's better to do it all in one go)

Thanks very much @adewit 
Sorry that it may be painful in this way.

**Comment by mmusich** [comment]
This can now be rebase for hopefully final testing

**Comment by srimanob** [comment]
@adewit 
Please add D87 workflow to 
https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_2026.py
So we can trigger the test of new geometry.

Thanks.

File: Configuration/PyReleaseValidation/python/relval_2026.py
Comment (adewit) [inline Configuration/PyReleaseValidation/python/relval_2026.py:39]: @srimanob D87 is added here - apparently this got mixed in during the rebase so it shows up as part of one of the earlier commits, sorry about that.

**Comment by srimanob** [comment]
test parameters:

**Comment by srimanob** [comment]
test parameters:
- workflows = 38634.0,39034.0

File: Configuration/PyReleaseValidation/python/relval_2026.py
Comment (srimanob) [reply Configuration/PyReleaseValidation/python/relval_2026.py:39]: Thanks, I missed it when I read the PR.

File: Configuration/Geometry/README.md
Comment (srimanob) [inline Configuration/Geometry/README.md:55]: Please clarify ig T25,26,27 supports DD4hep library as T24, or not (T21). The description of T26/T26 are unclear. Thanks.

**Comment by srimanob** [comment]
please test

File: Configuration/Geometry/README.md
Comment (adewit) [reply Configuration/Geometry/README.md:55]: Yes, they support the DD4Hep library as well (although all our internal tests have been done with DDD). I can update the README to clarify - I assume it won't interfere with the running tests in any way.

**Comment by srimanob** [comment]
Please test

**Comment by adewit** [comment]
Doesn't look like the error is related to these changes; need to test this with an IB that includes this: https://github.com/cms-sw/cmssw/pull/34532

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
I guess the tests will fail because of https://github.com/cms-sw/cmssw/issues/34552.

**Comment by cvuosalo** [comment]
PR #34534 would probably fix the Geometry unit tests that failed.

**Comment by tvami** [comment]
@cmsbuild , please test with  #34534

**Comment by yuanchao** [comment]
+1 

* adding run 2 tracker geo (T27) for bricked pixel
* unit tests and matrix tests passed except for 140.53 (different log files), 38634.0 and 39034.0 (missing map)
* SA warnings on AlignmentProducerBase.cc (esConsumes) and TrackerSystematicMisalignments.h (EDAnalyzer) are not of this PR modification.

**Comment by mmusich** [comment]
@cvuosalo @srimanob any further comment?

**Comment by cvuosalo** [comment]
+1

**Comment by srimanob** [comment]
+Upgrade

**Comment by mmusich** [comment]
@adewit this is now conflicting again because of `Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc`

**Comment by srimanob** [comment]
@cms-sw/db-l2 @cms-sw/pdmv-l2 @cms-sw/hlt-l2 Could you please consider this PR?

Since it touches several packages, so everytime small PR comes, it needs to rebase.

FYI @qliphy @perrotta Can we discuss the strategy of big PR in the ORP today. thanks.

**Comment by perrotta** [comment]
> @cms-sw/db-l2 @cms-sw/pdmv-l2 @cms-sw/hlt-l2 Could you please consider this PR?
> 
> Since it touches several packages, so everytime small PR comes, it needs to rebase.
> d 
> FYI @qliphy @perrotta Can we discuss the strategy of big PR in the ORP today. thanks.

We can discuss it, but I see no other possible strategy that either:
- marking this PR as "urgent", and then author and all L2s involved commit in modifying/evaluating this with the highest priority while ORP keeps all other overlapping PRs on hold during the (short) time this PR gets finalized
- or accepting that from time to time the PR will have to be rebased; once all signatures are collected we can move to the "urgent" case for the lastly rebased version, if needed

**Comment by Martin-Grunewald** [comment]
+1

**Comment by srimanob** [comment]
Urgent

We expect this to be available in per5. Rebase can be done at the last moment if need when we have all signatures. Thanks @perrotta

**Comment by srimanob** [comment]
Urgent

Need by pre5.

**Comment by bbilin** [comment]
+1

**Comment by civanch** [comment]
please test

**Comment by cvuosalo** [comment]
+1

**Comment by srimanob** [comment]
+Upgrade

**Comment by tvami** [comment]
+alca
 * but the PR needs to be rebased, so we'll have to sign again later anyway

**Comment by adewit** [comment]
I don't yet understand why the runTheMatrix tests are failing (can reproduce it locally though) - since it's a segfault in a module I don't know at all, I'll need time to figure out what is happening and fix it.

**Comment by tvami** [comment]
Hi @adewit, I think there are two ongoing things, 
(1) a timeout issue reported here https://github.com/cms-sw/cmssw/issues/34633 
and (2) that the PR https://github.com/cms-sw/cmssw/pull/34542 got merged which now leads to a merge conflict. 

Rebasing will help (2) and (1) is going to be handled by the core team. So please rebase and then we will trigger tests later on.

**Comment by qliphy** [comment]
please abort

**Comment by qliphy** [comment]
please test

**Comment by francescobrivio** [comment]
The `test_PixelBaryCentreTool` error was fixed in #34654

**Comment by francescobrivio** [comment]
@cmsbuild please test with #34654

**Comment by mmusich** [comment]
> and (2) that the PR #34542 got merged which now leads to a merge conflict.
 
  @tvami I don't see any merge conflict:
 
image

**Comment by srimanob** [comment]
For the recording, do you plan to fix the warning,
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7f0f25/17287/llvm-analysis/legacy-mod-sa.txt
in later PR? Sorry, it was mentioned somewhere in this long PR.

@adewit @mmusich

**Comment by srimanob** [comment]
@cms-sw/db-l2 Could you please consider and sign? Big thanks.

**Comment by ggovi** [comment]
+1

**Comment by qliphy** [comment]
+operations

**Comment by qliphy** [comment]
+1

**Comment by mmusich** [comment]
@srimanob 

> For the recording, do you plan to fix the warning,
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7f0f25/17287/llvm-analysis/legacy-mod-sa.txt
in later PR? Sorry, it was mentioned somewhere in this long PR.

Addressed at https://github.com/cms-sw/cmssw/pull/34690

**Pull request merged**
